cmd/ceremony/crl.go

```diff
@@ -62,3 +69,62 @@ func generateCRL(signer crypto.Signer, issuer *x509.Certificate, thisUpdate, nex
 
 	return pem.EncodeToMemory(&pem.Block{Type: "X509 CRL", Bytes: crlBytes}), nil
 }
+
+// truncatedHash computes a truncated SHA1 hash across arbitrary bytes. Uses
+// SHA1 because that is the algorithm most commonly used in OCSP requests.
+// PURPOSEFULLY NOT EXPORTED. Exists only to ensure that the implementations of
+// Certificate.NameID() and GetIssuerNameID() never diverge. Use those instead.
+func truncatedHash(name []byte) issuerNameID {
+	h := crypto.SHA1.New()
+	h.Write(name)
+	s := h.Sum(nil)
+	return issuerNameID(big.NewInt(0).SetBytes(s[:7]).Int64())
+}
+
+// distributionPointName represents the ASN.1 DistributionPointName CHOICE as
+// defined in RFC 5280 Section 4.2.1.13. We only use one of the fields, so the
+// others are omitted.
+type distributionPointName struct {
+	// Technically, FullName is of type GeneralNames, which is of type SEQUENCE OF
+	// GeneralName. But GeneralName itself is of type CHOICE, and the ans1.Marhsal
+	// function doesn't support marshalling structs to CHOICEs, so we have to use
+	// asn1.RawValue and encode the GeneralName ourselves.
+	FullName []asn1.RawValue `asn1:"optional,tag:0"`
+}
+
+// issuingDistributionPoint represents the ASN.1 IssuingDistributionPoint
+// SEQUENCE as defined in RFC 5280 Section 5.2.5. We only use two of the fields,
+// so the others are omitted.
+type issuingDistributionPoint struct {
+	DistributionPoint   distributionPointName `asn1:"optional,tag:0"`
+	OnlyContainsCACerts bool                  `asn1:"optional,tag:2"`
+}
+
+// makeIDPExt returns a critical IssuingDistributionPoint extension containing a
+// URI built from the base url and the issuer's NameID. It also sets the
+// OnlyContainsCACerts boolean to true.
+func makeIDPExt(base string, issuer issuerNameID) (*pkix.Extension, error) {
```

Comment:
This is not how we want to compute the IDP URL here. The IDP URL should be taken directly from the config file and not be modified or templated at all.

Recall that these CRLs are fundamentally different from the CRLs which contain our end-entity certificates. We have a lot of end-entity CRLs, so we don't want to manage them manually. Instead we name them programmatically, using the intermediate's IssuerNameID and the CRL's ShardIdx in the URL. But these CRLs are issued directly from our Roots, of which we have very few. And those roots have already issued Intermediates which _say where their CRL can be found_:
![image](https://github.com/letsencrypt/boulder/assets/2474926/575c7295-2355-46c5-985d-0cd0da8bef23)

The URL here, inside this CRL which may one day revoke that intermediate, has to _exactly match_ the URL contained inside that intermediate. No ShardIdx shenanigans (which you've already removed), and no IssuerNameID shenanigans either.

So we still want a function like this, but it should only take the URL itself as input. That lets you remove the `issuerNameID` type and the `truncatedHash` function.

And a potential lint-ish check would be to ensure that all of the intermediates being revoked in this CRL have a byte-for-byte identical crlDistributionPoint.distributionPoint as this issuingDistributuionPoint.distributionPoint.

-----

_Alternatively_, the `distributionPoint` field could be left out of this extension entirely:
> RFC 5280, Section 5.2.5: If the distributionPoint field is present, then it MUST include at least one of names from the corresponding distributionPoint field of the cRLDistributionPoints extension of every certificate that is within the scope of this CRL. The identical encoding MUST be used in the distributionPoint fields of the certificate and the CRL.
>
> If the distributionPoint field is absent, the CRL MUST contain entries for all revoked unexpired certificates issued by the CRL issuer, if any, within the scope of the CRL.

So since we're not doing sharded CRLs at this level, we can just leave the `distributionPoint` field out entirely. Then we don't have to worry about ensuring that it matches what we've put in our intermediates (allowing us to change that crlDistributionPoint URL over time), and we don't have to worry about encoding it here at all. Just put in the `onlyContainsCACerts: true` boolean and nothing else.

Reply:
Ah, thank you for the thorough explanation. I'm going to choose the latter path.